intel_extension_for_transformers/backends/neural_engine/graph/application/ChatOPT/main_opt.cpp

```diff
@@ -0,0 +1,652 @@
+        }
+
+        if (temp <= 0) {
+          // Greedy sampling
+          id = model_sample_token_greedy(ctx, &candidates_p);
+        } else {
+          if (mirostat == 1) {
+            static float mirostat_mu = 2.0f * mirostat_tau;
+            const int mirostat_m = 100;
+            model_sample_temperature(ctx, &candidates_p, temp);
+            id = model_sample_token_mirostat(ctx, &candidates_p, mirostat_tau, mirostat_eta, mirostat_m, &mirostat_mu);
+          } else if (mirostat == 2) {
+            static float mirostat_mu = 2.0f * mirostat_tau;
+            model_sample_temperature(ctx, &candidates_p, temp);
+            id = model_sample_token_mirostat_v2(ctx, &candidates_p, mirostat_tau, mirostat_eta, &mirostat_mu);
+          } else {
+            // Temperature sampling
+            model_sample_top_k(ctx, &candidates_p, top_k, 1);
+            model_sample_tail_free(ctx, &candidates_p, tfs_z, 1);
+            model_sample_typical(ctx, &candidates_p, typical_p, 1);
+            model_sample_top_p(ctx, &candidates_p, top_p, 1);
+            model_sample_temperature(ctx, &candidates_p, temp);
+            id = model_sample_token(ctx, &candidates_p);
+          }
+        }
+        // printf("`%d`", candidates_p.size);
+
+        if (embd.size() > 0 && !path_session.empty()) {
+          session_tokens.insert(session_tokens.end(), embd.begin(), embd.end());
+          n_session_consumed = session_tokens.size();
+        }
+      }
+
+      embd.clear();
+
+      if ((int)embd_inp.size() <= n_consumed && !is_interacting) {
+        // optionally save the session on first sample (for faster prompt loading next time)
+        if (!path_session.empty() && need_to_save_session) {
+          need_to_save_session = false;
+          model_save_session_file(ctx, path_session.c_str(), session_tokens.data(), session_tokens.size());
+        }
+
+        last_n_tokens.erase(last_n_tokens.begin());
+        last_n_tokens.push_back(id);
+      }
+
+      // replace end of text token with newline token when in interactive mode
+      if (id == model_token_eos() && params.interactive && !params.instruct) {
+        id = model_token_newline.front();
+        if (params.antiprompt.size() != 0) {
+          // tokenize and inject first reverse prompt
+          const auto first_antiprompt = ::model_tokenize(ctx, params.antiprompt.front(), false);
+          embd_inp.insert(embd_inp.end(), first_antiprompt.begin(), first_antiprompt.end());
+        }
+      }
+
+      // add it to the context
+      embd.push_back(id);
+
+      // echo this to console
+      input_echo = true;
+
+      // decrement remaining sampling budget
+      --n_remain;
+    } else {
+      // some user input remains from prompt or interaction, forward it to processing
+      // OPT model use GPT-2 tokenizer. It will prepend bos_token when encode user's prompt
+      // https://github.com/huggingface/transformers/issues/3311
+      embd.push_back(OPT_BOS_TOKEN);
```

Comment:
@intellinjun Can you help investigate how to merge the `prepend bos_token_id` action into [`model_tokenize`](https://github.com/intel/intel-extension-for-transformers/blob/opt_graph/intel_extension_for_transformers/backends/neural_engine/graph/models/opt/opt.cpp#L409-L417) function by set `add_bos = true`?

Reply:
sync with @airMeng, please add a member `int bos_token_id =-1` (-1 means unknown) in `model_vocab`. And use `model_vocab` to prepend `eos` token in `input_ids` if needed.